File: python/tater/optimizer.py
from __future__ import annotations

import math
from typing import Sequence

from .tensor import Tensor, clip_grad_norm, zero_grad


class Adam:
    def __init__(
        self,
        params: Sequence[Tensor],
        learning_rate: float = 0.001,
        beta1: float = 0.9,
        beta2: float = 0.999,
        epsilon: float = 1e-8,
    ) -> None:
        if learning_rate <= 0.0:
            raise ValueError("Adam learning rate must be positive")
        self.params = list(params)
        self.learning_rate = learning_rate
        self.beta1 = beta1
        self.beta2 = beta2
        self.epsilon = epsilon
        self.t = 0
        self.m: list[list[float]] = []
        self.v: list[list[float]] = []
        for param in self.params:
            if param is None or not param.defined:
                raise ValueError("Adam received an undefined parameter")
            self.m.append([0.0] * param.size)
            self.v.append([0.0] * param.size)

    def zero_grad(self) -> None:
        zero_grad(self.params)

    def step(self, clip_norm: float = 1.0) -> float:
        unclipped_norm = clip_grad_norm(self.params, clip_norm)
        self.t += 1

        bias1 = 1.0 - self.beta1**self.t
        bias2 = 1.0 - self.beta2**self.t

        for p, param in enumerate(self.params):
            for i in range(param.size):
                grad = param.grad[i]
                self.m[p][i] = self.beta1 * self.m[p][i] + (1.0 - self.beta1) * grad
                self.v[p][i] = self.beta2 * self.v[p][i] + (1.0 - self.beta2) * grad * grad
                m_hat = self.m[p][i] / bias1
                v_hat = self.v[p][i] / bias2
                param.data[i] -= self.learning_rate * m_hat / (
                    math.sqrt(v_hat) + self.epsilon
                )

        return unclipped_norm

File: cpp/src/optimizer.cpp
#include "tater/optimizer.hpp"

#include <cmath>
#include <stdexcept>
#include <utility>

namespace tater {

Adam::Adam(std::vector<Tensor*> params,
           double learning_rate,
           double beta1,
           double beta2,
           double epsilon)
    : params_(std::move(params)),
      learning_rate_(learning_rate),
      beta1_(beta1),
      beta2_(beta2),
      epsilon_(epsilon) {
    if (learning_rate_ <= 0.0) {
        throw std::runtime_error("Adam learning rate must be positive");
    }
    for (Tensor* param : params_) {
        if (param == nullptr || !param->defined()) {
            throw std::runtime_error("Adam received an undefined parameter");
        }
        m_.emplace_back(param->size(), 0.0);
        v_.emplace_back(param->size(), 0.0);
    }
}

void Adam::zero_grad() {
    tater::zero_grad(params_);
}

double Adam::step(double clip_norm) {
    const double unclipped_norm = clip_grad_norm(params_, clip_norm);
    ++t_;

    const double bias1 = 1.0 - std::pow(beta1_, static_cast<double>(t_));
    const double bias2 = 1.0 - std::pow(beta2_, static_cast<double>(t_));

    for (std::size_t p = 0; p < params_.size(); ++p) {
        Tensor& param = *params_[p];
        for (std::size_t i = 0; i < param.size(); ++i) {
            const double g = param.grad()[i];
            m_[p][i] = beta1_ * m_[p][i] + (1.0 - beta1_) * g;
            v_[p][i] = beta2_ * v_[p][i] + (1.0 - beta2_) * g * g;
            const double m_hat = m_[p][i] / bias1;
            const double v_hat = v_[p][i] / bias2;
            param.data()[i] -= learning_rate_ * m_hat / (std::sqrt(v_hat) + epsilon_);
        }
    }

    return unclipped_norm;
}

} // namespace tater
